### PR TITLE
executor: fix inconsistent of grants privileges with MySQL when executing `grant all on ...`

### PR DESCRIPTION
**Pull request by tsthght**
### What problem does this PR solve? 
1 fix bug for #12324
2 fix bug for execute SQL `show grants for xxx`

### What is changed and how it works?
create user "ght1";
grant all on . to 'ght1'@'%' identified by 'Poiuy123';
select * from mysql.user where user='ght1'\G
*************************** 1. row ***************************
Grant_priv: N

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by imtbkcat** [comment]
@tsthght Please fix test below:
```
privileges_test.go:265
privileges_test.go:504
privileges_test.go:694
```
And add some test to test `GRANT ALL` and `GRANT ... WITH GRANT OPTION`

File: executor/grant.go [hunk 1]
```diff
@@ -303,6 +303,9 @@ func composeGlobalPrivUpdate(priv mysql.PrivilegeType, value string) (string, er
 	if priv == mysql.AllPriv {
 		strs := make([]string, 0, len(mysql.Priv2UserCol))
 		for _, v := range mysql.Priv2UserCol {
+			if v == mysql.Priv2UserCol[mysql.GrantPriv] {
```
[the remark this answers is not in the document]
Comment (tsthght) [reply executor/grant.go]: indeed, I will try it.

**Comment by tiancaiamao** [comment]
Any update? @tsthght

**Comment by tsthght** [comment]
> Any update? @tsthght

It will be updated today.

File: privilege/privileges/privileges_test.go [hunk 2]
```diff
@@ -264,6 +264,20 @@ func (s *testPrivilegeSuite) TestShowGrants(c *C) {
 	c.Assert(gs, HasLen, 1)
 	c.Assert(gs[0], Equals, `GRANT ALL PRIVILEGES ON *.* TO 'show'@'localhost'`)
 
+	// All privileges with grant option
+	mustExec(c, se, `GRANT ALL ON *.* TO 'show'@'localhost' WITH GRANT OPTION;`)
+	gs, err = pc.ShowGrants(se, &auth.UserIdentity{Username: "show", Hostname: "localhost"}, nil)
+	c.Assert(err, IsNil)
+	c.Assert(gs, HasLen, 1)
+	c.Assert(gs[0], Equals, `GRANT ALL PRIVILEGES ON *.* TO 'show'@'localhost' WITH GRANT OPTION`)
+
```
Comment (tiancaiamao) [inline privilege/privileges/privileges_test.go]: And also check the GrantPriv column is `Y`?

File: privilege/privileges/privileges_test.go [hunk 3]
```diff
@@ -264,6 +264,20 @@ func (s *testPrivilegeSuite) TestShowGrants(c *C) {
...
+	// Revoke grant option
+	mustExec(c, se, `REVOKE GRANT OPTION ON *.* FROM 'show'@'localhost';`)
+	gs, err = pc.ShowGrants(se, &auth.UserIdentity{Username: "show", Hostname: "localhost"}, nil)
+	c.Assert(err, IsNil)
+	c.Assert(gs, HasLen, 1)
+	c.Assert(gs[0], Equals, `GRANT ALL PRIVILEGES ON *.* TO 'show'@'localhost'`)
+
```
Comment (tiancaiamao) [inline privilege/privileges/privileges_test.go]: And check the GrantPriv column is `N` ?

**Comment by tiancaiamao** [comment]
You can use `go mod edit github.com/pingcap/parser=github.com/tsthght/parser@pri` to use the right parser version. @tsthght

**Comment by tsthght** [comment]
> You can use `go mod edit github.com/pingcap/parser=github.com/tsthght/parser@pri` to use the right parser version. @tsthght

Great! Thank you, I will try it~

File: privilege/privileges/privileges_test.go (hunk 3 above)
Comment (tsthght) [reply privilege/privileges/privileges_test.go]: I Had added more testcase, thank you~ 
image

File: go.mod [hunk 4]
```diff
@@ -79,3 +80,5 @@ require (
 )
 
 go 1.13
+
+replace github.com/pingcap/parser => github.com/tsthght/parser-1 v0.0.0-20191017013159-be204595253e
```
Comment (tiancaiamao) [inline go.mod]: You need to remove this line, and use 

```
go get -u github.com/pingcap/parser@master
```

Comment (tiancaiamao) [reply go.mod]: Rest LGTM

Comment (tiancaiamao) [reply go.mod]: It seems you still have to call `go mod tidy`...

**Comment by tiancaiamao** [comment]
LGTM @imtbkcat

**Comment by tiancaiamao** [comment]
```

[2019-10-21T13:31:06.460Z] "SHOW GRANTS FOR current_user() USING `engineering`@`%`,`role`@`%`;" 

[2019-10-21T13:31:06.460Z]  around line 114, 

[2019-10-21T13:31:06.460Z] we need(221):

[2019-10-21T13:31:06.460Z] SHOW GRANTS FOR current_user() USING `engineering`@`%`,`role`@`%`;

[2019-10-21T13:31:06.460Z] Grants for root@127.0.0.1

[2019-10-21T13:31:06.460Z] GRANT ALL PRIVILEGES ON *.* TO 'root'@'%'

[2019-10-21T13:31:06.460Z] GRANT 'engineering'@'%', 'role'@'%' TO 'root'@'%'

[2019-10-21T13:31:06.460Z] GRANT role1 TO current_user();

[2019-10-21T13:31:06.460Z] SET R

[2019-10-21T13:31:06.460Z] but got(221):

[2019-10-21T13:31:06.460Z] SHOW GRANTS FOR current_user() USING `engineering`@`%`,`role`@`%`;

[2019-10-21T13:31:06.460Z] Grants for root@127.0.0.1

[2019-10-21T13:31:06.460Z] GRANT ALL PRIVILEGES ON *.* TO 'root'@'%' WITH GRANT OPTION

[2019-10-21T13:31:06.460Z] GRANT 'engineering'@'%', 'role'@'%' TO 'root'@'%' WITH GRANT OPTION

[2019-10-21T13:31:06.460Z] 
```

**Comment by imtbkcat** [comment]
I will fix mysql sql test.

File: privilege/privileges/cache.go [hunk 5]
```diff
@@ -867,9 +942,15 @@ func (p *MySQLPrivilege) showGrants(user, host string, roles []*auth.RoleIdentit
 				g += ", "
 			}
 		}
-		s := fmt.Sprintf(`GRANT %s TO '%s'@'%s'`, g, user, host)
+		var s string
+		if hasGrantOptionPriv {
```
[the remark this answers is not in the document]
Comment (tsthght) [reply privilege/privileges/cache.go]: OK, I will remove it.

**Comment by imtbkcat** [comment]
I will approve after test merge, wait a moment.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**